src/rest.cpp

```diff
@@ -644,7 +667,9 @@ static bool rest_blockhash_by_height(const std::any& context, HTTPRequest* req,
 
     CBlockIndex* pblockindex = nullptr;
     {
-        ChainstateManager& chainman = EnsureAnyChainman(context);
+        ChainstateManager* maybe_chainman = GetChainman(context, req);
```

Comment:
In commit "rest: Add GetChainman function and use it" (a839078c20f4ac3d92e14af7b287f2f0fddbc61e)

re: "Someone can probably make this a bit prettier," your current version of this seems fine to me, but if you wanted to make it more minimal you could use CHECK_NONFATAL and catch the NonFatalCheckError exception in here, I think: https://github.com/bitcoin/bitcoin/blob/06d573f053c63eb6521c62d210c5924418ae2b83/src/rest.cpp#L699

Reply:
re: https://github.com/bitcoin/bitcoin/pull/21767#discussion_r628449195

> Hmmmm, how could we differentiate between NOT_FOUND like the "no mempool"-case and INTERNAL_SERVER_ERROR like the "no Context/Chainman"-case here?

I didn't realize that was intentional and I do think the your approach is fine, but you wanted to use the NonFatalCheckError approach, you would just add:

```
class PageNotFoundError : public std::runtime_error
{
    using std::runtime_error::runtime_error;
};

...

throw PageNotFoundError{};

...

} catch (const PageNotFoundError&) {

```

Reply:
Oh I see see the mempool code you're referring to is preexisting. That would more incline towards following previous pattern like you're doing I think. I'm sure you could just set the 404 one way and the 500 the other way, too though. Again any approach seems fine and I was just responding to dissatisfaction in the commit message.